packages/SwingSet/src/kernel/kernel.js

```diff
@@ -396,42 +395,70 @@ export default function buildKernel(kernelEndowments) {
     }
   }
 
-  function addGenesisVat(name, setup, options = {}) {
-    name = `${name}`;
-    // for now, we guard against 'options' by treating it as JSON-able data
-    options = JSON.parse(JSON.stringify(options));
-    // todo: consider having vats indicate 'enablePipelining' during setup(),
-    // rather than using options= during kernel.addGenesisVat()
+  // this is for unit tests
+  function addGenesisVatSetup(name, setup, options = {}) {
+    if (typeof setup !== 'function') {
+      throw Error(`setup is not a function, rather ${setup}`);
+    }
     const knownOptions = new Set(['enablePipelining']);
     for (const k of Object.getOwnPropertyNames(options)) {
       if (!knownOptions.has(k)) {
-        throw new Error(`unknown option ${k}`);
+        throw Error(`unknown option ${k}`);
       }
     }
 
     if (started) {
-      throw new Error(`addGenesisVat() cannot be called after kernel.start`);
+      throw Error(`addGenesisVat() cannot be called after kernel.start`);
     }
     if (genesisVats.has(name)) {
-      throw new Error(`vatID ${name} already added`);
+      throw Error(`vatID ${name} already added`);
     }
 
     genesisVats.set(name, { setup, options });
   }
 
-  function addGenesisDevice(name, buildRootDeviceNode, endowments) {
+  function addGenesisVat(name, bundle, options = {}) {
+    // todo: consider having vats indicate 'enablePipelining' by exporting a
```

Comment:
To what extent is pipelineability an innate property of a vat (i.e., it's in the bundle) versus an expression of the intent of whoever is instantiating the vat?  In other words, should vats indicate not 'enablePipelining' but instead 'pipeliningCapable', and then have an 'enablePipelining' creation option expressing the creator's desire to have pipelining turned on, with the actual enablement being the logical AND of the two.  Moreover, it seems to me that potentially it's also partially a function of the environment into which the vat is being loaded -- i.e., a low-memory substrate may prefer senders to buffer messages whereas a high-memory substrate may prefer to be more aggressive.

Reply:
Indeed, #1354 is about making it clear that pipelineability is an innate property of the vat: if the vat isn't prepared to accept messages without a concrete fulfilled-object target (which means forwarding them to someone else, or queueing them for later), then it shouldn't receive them. So setting `options.enablePipelining` on a vat that doesn't want it is a bug.

In the other direction, if a vat can handle pipelining, but for whatever reason the kernel decides to not use it, the consequence is probably just additional latency. I can't think of any good reason to do this: queueing those messages in the kernel just consumes memory on things that could have been delivered instead. If the communication link is down, the send-to-remote messages will probably just queue up somewhere else.

So I think either vats should indicate they are capable and the kernel should honor that, or they should indicate they are *not* capable and the kernel should honor that too. My plan with #1354 is to remove the kernel-side config and rely purely upon the vat's self-reported wishes.

Reply:
Upon further reflection, the other issue I raised about the suitability of an instantiation environment to pipelining is really a swingset-level concern, i.e., if a vat is in a low memory environment, it's because it's running on a swingset in a low memory environment, so the selectivity of whether or not to pipeline a specific message send in that case is really about the relationship between the sending and receiving swingsets, i.e., if the sending and receiving vats are running in different swingsets on (likely) different machines they will be linked by a vatTP connection, and the control over the enablement of pipelining there should be part of the configuration of that connection, not of the vats.

Reply:
Another thought: I could imagine a vat creator wanting to selectively turn off pipelining as part of testing.  Such control shouldn't be edge case tail that wags the API design dog, but it seems plausible that we might end up wanting some kind of kernel testing back door for this, analogous to what you're doing with `createRootObject` vs. `setup`.

Reply:
yeah, I ran into that while experimenting with #1354.. `test-message-patterns.js` has cases which expect different results when pipelining is enabled/disabled, which I needed when first implementing pipelining. Since it was switched on, the other half of those `expected` checks are basically commented out. But at this point, if we add any explicit tests for pipelining, I think we'd stick to using the exported boolean to control it, rather than trying to configure it externally.